### ISSUE TITLE
[7.2.1] `http_archive` not downloading files when using a private bazel registry.

### ISSUE THREAD

**Issue by Jannick52** [bazelbuild/bazel#23266]
### summary
I am having a problem where an external dependency (a bazel module) that I am referencing through a bazel registry is not downloading the files in the `http_archive` rule that is defined in the `MODULE.bazel` file of that dependency.

### description
I have a bazel module `my_module` (in which I call `bazel build`) that is referencing an external dependency (module) `wrapper-mavsdk` through a private bazel registry. The external dependency has a `http_archive` rule defined in its `MODULE.bazel` file that is supposed to download the `mavsdk` files from the public mavsdk repository. The `mavsdk` module is then built using the `cmake` rule from `foreign_cc_rules`.
When I leave out the bazel registry and directly reference the `wrapper-mavsdk` module from `my_module` this works fine, and the `mavsdk` files are downloaded and the `mavsdk` target is built corrrectly:

```
bazel_dep(name = "wrapper-mavsdk", version = "1.0.0")
git_override(
    module_name = "wrapper-mavsdk",
    remote = "https://github.com/org/wrapper-library",
    commit = "52j3138c39eb8ddl34l3j543J27cbad20eca473ac",
    strip_prefix = "mavsdk"
)
```

But now I want to use a private bazel registry so that I can remove the `git_override` rule and simply reference `wrapper-mavsdk` through `bazel_dep`. After implementing the registry and running the build again, I can see that the `wrapper-mavsdk` module is correctly being loaded through the registry since in the error (shown below) the module's BUILD.bazel file is being mentioned and I checked that the sandbox contains the correct contents for the BUILD.bazel file. The problem is that the `cmake` rule is failing because the `mavsdk` files that were supposed to be downloaded by the `http_archive` rule in the wrapper-mavsdk `MODULE.bazel` are not present. Here is the error:

```
jovyan@jannick52-laptop1:~/bprt-dam-c-src/bprt-dam-c-src$ bazel build //...
INFO: Invocation ID: 87d81608-5fe6-43a3-8250-e9e9a6e75aa2
INFO: Analyzed 11 targets (0 packages loaded, 0 targets configured).
ERROR: /home/jovyan/.cache/bazel/_bazel_jovyan/a8899711855f9ee8e459498dd22dccac/external/wrapper-mavsdk~/BUILD.bazel:28:6: Foreign Cc - CMake: Building mavsdk failed: (Exit 1): bash failed: error executing CcCmakeMakeRule command (from target @@wrapper-mavsdk~//:mavsdk) 
  (cd /home/jovyan/.cache/bazel/_bazel_jovyan/a8899711855f9ee8e459498dd22dccac/sandbox/linux-sandbox/27/execroot/_main && \
  exec env - \
    CARGO_HOME=/home/jovyan/.cargo \
    CMAKE_BUILD_PARALLEL_LEVEL=8 \
    PATH=/home/jovyan/.cache/bazelisk/downloads/sha256/80ccd1ecb4b88750fbe5d7622d67072fddcba9da7808f13356555e480bf67875/bin:/home/jovyan/.cargo/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
    RUSTUP_HOME=/home/jovyan/.rustup \
  /bin/bash -c bazel-out/k8-fastbuild/bin/external/wrapper-mavsdk~/mavsdk_foreign_cc/wrapper_build_script.sh)
# Configuration: 16b72eaf37fb94883d69ac6c76e64e2c124f672d2690b164d7f817b5a1a6a159
# Execution platform: @@platforms//host:host
```

Here are all of the relevant code files:

### my_module

my_module `MODULE.bazel`:
```
module(name = "my_module")

bazel_dep(name = "wrapper-mavsdk", version = "1.0.0")
```

my_module `.bazelrc`:
```
build --registry=https://bcr.bazel.build
build --registry=https://raw.githubusercontent.com/Jannick52/my-bazel-registry/main
```

### wrapper-mavsdk external module
wrapper-mavsdk `MODULE.bazel` (same as in bazel registry wrapper-mavsdk 'MODULE.bazel`):
```
module(name = "wrapper-mavsdk", version = "1.0.0")

bazel_dep(name = "rules_foreign_cc", version = "0.10.1")

http_archive = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "mavsdk",
    url = "https://github.com/mavlink/MAVSDK/archive/refs/tags/v2.10.0.tar.gz",
    build_file = ":BUILD.bazel",
    strip_prefix = "MAVSDK-2.10.0",
)
```

wrapper-mavsdk `BUILD.bazel`:
```
load("@rules_foreign_cc//foreign_cc:defs.bzl", "cmake")

cmake(
    name = "mavsdk",
    cache_entries = {
        "BUILD_MAVSDK_SERVER": "OFF",
        "BUILD_TESTS": "OFF",
        "CMAKE_BUILD_TYPE": "Release",
        "DBUILD_SHARED_LIBS:": "OFF",
    },
    lib_source = ":all_srcs",
    out_include_dir = "include/mavsdk",
    # The cmake rule is producing libmavsdk.so, libmavsdk.so.2, and libmavsdk.so.0.0.0.
    # libmavsdk.so.0.0.0 is the actual shared library, and the other two are symbolic links.
    out_shared_libs = [
        "libmavsdk.so",
        "libmavsdk.so.0.0.0",
    ],
    tags = ["requires-network"],
    visibility = ["//visibility:public"],
)

# Selects all files that should be downloaded by the http_archive rule
filegroup(
    name = "all_srcs",
    srcs = glob(["**"]),
    visibility = ["//visibility:public"],
)
```

### private bazel registry containing wrapper-mavsdk entry

bazel registry wrapper-mavsdk `source.json`:
```
{
    "type": "git_repository",
    "remote": "https://github.com/org/wrapper-library",
    "commit": "b13e45db653567b184845a4613ace1e26414d391",
    "strip_prefix": "mavsdk",
    "verbose": "true"
}
```
bazel registry wrapper-mavsdk 'MODULE.bazel` (same as in wrapper-mavsdk `MODULE.bazel`):
```
module(name = "wrapper-mavsdk", version = "1.0.0")

bazel_dep(name = "rules_foreign_cc", version = "0.10.1")

http_archive = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "mavsdk",
    url = "https://github.com/mavlink/MAVSDK/archive/refs/tags/v2.10.0.tar.gz",
    build_file = ":BUILD.bazel",
    strip_prefix = "MAVSDK-2.10.0",
)
```

I could not figure out why the `http_archive` rule was not being called when the wrapper-mavsdk module was being pulled in through the bazel registry. Hopefully someone can provide some guidance on this issue. Thank you kindly for taking the time to read my issue.

### Which category does this issue belong to?

External Dependency

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

See bug description.

### Which operating system are you running Bazel on?

Ubuntu 22.04.3 LTS

### What is the output of `bazel info release`?

release 7.2.1

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text
-
```

### Have you found anything relevant by searching the web?

Unfortunately I could not find anything relevant in the Github issues or anywhere else.

**Comment by meteorcloudy** [comment]
I believe this is because
```
http_archive(
    name = "mavsdk",
    url = "https://github.com/mavlink/MAVSDK/archive/refs/tags/v2.10.0.tar.gz",
    build_file = ":BUILD.bazel",
    strip_prefix = "MAVSDK-2.10.0",
)
``` 
isn't defined in "bazel registry wrapper-mavsdk 'MODULE.bazel'"

The MODULE.bazel file checked in the registry should be the same as the one fetched by the source.json file. Otherwise, the one in the Bazel registry takes priority.

**Comment by Jannick52** [comment]
@meteorcloudy Thank you for the response. I realized this as well and updated "bazel registry wrapper-mavsdk `MODULE.bazel`" to reflect the `MODULE.bazel` file in the source.json file. Unfortunately however this did not solve the issue. The mavsdk files are still not being downloaded with the `http_archive` rule. I just confirmed again that this is the case.

I will update my original issue to include this change.

**Comment by meteorcloudy** [comment]
After you change the registry content, can you please run with `--lockfile_mode=refresh`? Otherwise, the old content might be cached by Bazel since Bazel by default assumes the registry is immutable.

**Comment by Jannick52** [comment]
@meteorcloudy Unfortunately that still gives the same results. I have also tried it with a completely fresh install and that also did not help.
Looking at the [picotool MODULE.bazel](https://github.com/bazelbuild/bazel-central-registry/blob/842b622ca8bc5bfe02f1ad251f3152123f9a0a2c/modules/picotool/2.0.0/MODULE.bazel#L19-L26) in the bazel central registry it does seem that `http_archive` works normally with a registry. I wonder if it has to do with the `foreign_cc_rules` `cmake` rule.

**Comment by meteorcloudy** [comment]
Can you please try to assemble everything in a git repo and provide it as a minimal repro case? I'm pretty sure there is something wrong with the setup, this functionality should just work.

**Comment by Jannick52** [comment]
I created a minimal repro case and indeed everything worked as expected. I then went back to my original setup and this is now also working. There must indeed have been an issue with my setup but unfortunately I am not sure what caused the difference. I still have issues using the bazel registry if it is in a private github repository (`module not found in registries: wrapper-mavsdk@1.0.0`) but that is a separate issue.

Thank you for the help! I will list the repositories for the working minimal example in case anyone else ever needs it:

- main module: https://github.com/Jannick52/bazel_issue_23266_example
- registry: https://github.com/Jannick52/bzl-registry
- wrapper library: https://github.com/Jannick52/3rd-party-libs
